### PR TITLE
feat(cta-block): introduced cta block web component, added it as cta section variant, removed promo group and promo item

### PR DESCRIPTION
**Pull request by rastogir**
### Related Ticket(s)

- [#5216](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5216)
- [#5220](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5220)
- [#5221](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5221)
- [#5218](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5218)

### Description

- Adds `<dds-cta-block>`. This replaces `<dds-promo-group>` and `<dds-promo-item>`. It is based off of `<dds-content-item>`.
- Adds `CtaBlock` variant to `<dds-cta-section>` component. This variant has an additional prop of `Media Type` to switch between image, video, statistic and pictogram. 
- Removes `<dds-promo-group>` and `<dds-promo-item>`.

### Design Specs
 
- https://ibm.ent.box.com/s/shr0pnhof0yzk2b5lh0nokwopgxxb9rr
Screen Shot 2021-03-17 at 3 54 43 PM

- Additional design spec for statisitc from Kirby :
  - Font: IBM Plex Sans, SemiBold
  - Max: 76px
  - xlg: 60px
  - lg: 54px
  - md and sm: 42px

### Functional Spec

- https://user-images.githubusercontent.com/3793636/108720924-067c2a00-74e7-11eb-8441-44bb8b412dcd.png

### Changelog

**New**

- `<dds-cta-block>` component. 

**Changed**

- `<dds-cta-section>` story updated with `<dds-cta-block>` variant.

**Removed**

- `<dds-promo-group>` and `<dds-promo-item>` component.

### CODE REVIEW

File: packages/web-components/src/components/cta-block/__stories__/cta-block.stories.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+          <dds-cta-block-item>
+            ${mediaType === 'image' ? image : ``} ${mediaType === 'video' ? video : ``}
+            ${mediaType === 'pictogram' ? pictogram : ``} ${mediaType === 'statistic' ? statistic : ``}
+            <dds-content-item-heading>${itemHeading}</dds-content-item-heading>
+            <dds-content-item-copy>${itemCopy}</dds-content-item-copy>
+            <dds-link-with-icon href="${linkWithIcon.href}" slot="footer">
+              ${linkWithIcon.copy} ${ArrowRight20({ slot: 'icon' })}
+            </dds-link-with-icon>
+          </dds-cta-block-item>
+        `
```
Comment (asudoh) [inline packages/web-components/src/components/cta-block/__stories__/cta-block.stories.ts:147]: I think render-from-data approach complicates the explanation of how to use the component, would you put plain markup here?

**Review by asudoh — COMMENTED** [review]
Looks a nice improvement. Do you want to create a CodeSandbox example? Thanks @rastogir!

**Review by photodow — CHANGES_REQUESTED** [review]
Thanks @rastogir! Here are some quick high-level things I noticed... Reach out if you have any questions.

1. `CTA block`, and `CTA section` are backwards. What you have as `CTA block` should be `CTA section` and vice versa.
image

2. Casing needs to be fixed. We use sentence casing for these.
```
CTA block
— Default
CTA section
— Default
— With content items
— With link list
— CTA block
```

3. Is `CTA section` missing bottom spacing?

image

4. The horizontal rules are using two different colors. I don't think that's intended, right?

image

5. Looks like some spacing issues with the supporting items.

image

6. Nitpicky thing, but is this a `media type`? The functional specs I call it a `supporting element`. I'm not necessarily saying we HAVE to use that here, but I don't think `media type` is accurate.

image

7. I think you have the video title and duration in the wrong place.

image

8. Why does the type size change here? CTA block should be using expressive-heading-05 for the heading, and `heading-03` for that copy.

image

image

**Comment by kirbyevans** [comment]
Hey, @rastogir here are some issues I'm noticing in addition to @photodow 's notes from above.

1.Content section heading in the left column area should only span 3 columns
2. Main content area needs shifted left to align with the 16-column grid.
 
Pasted_Image_3_18_21__4_40_PM

----

1. Appears to be some alignment issues with the Content item component. Additional space getting added to title on shorter Content items.
2. CTA text link with icon should consistently align based on longest content item character count. See design spec for more detail.
3. I want to also confirm that an additional CTA text link with icon can be added to a Content item for a max of two links if needed.

Pasted_Image_3_18_21__4_55_PM

----

@photodow did we determine if this logic was in scope or not?

> **Minimum of two items per row**
> Avoid orphaned rows with one item. When this occurs the item in the third column from the row above should drop to the row below.

----

1. Bottom spacing for items in the last row should be 160px.

Pasted_Image_3_18_21__5_21_PM

**Comment by photodow** [comment]
> @photodow did we determine if this logic was in scope or not?

Out of scope. There should be a `# of column` prop so the AEM CMS can drop it if a given orphan is present. I didn't see a knob to test this, but I did see one story had 2 columns, and another story had 3 columns.

**Comment by rastogir** [comment]
@asudoh 
- [ ]  Created sandbox example however I'll need to update it before PR is merged 
- [ ] Also added plain markup 

@photodow 
- [ ] CTA block, and CTA section are backwards. What you have as CTA block should be CTA section and vice versa. 
- [ ] Casing needs to be fixed. We use sentence casing for these

  - Meeting scheduled for 3/22 to discuss the requirement and implementation for above two feedbacks.

- [ ] Is CTA section missing bottom spacing?
- [ ] The horizontal rules are using two different colors. I don't think that's intended, right? _This is as per design spec_  
- [ ] Looks like some spacing issues with the supporting items.
- [ ] Nitpicky thing, but is this a media type? The functional specs I call it a supporting element. I'm not necessarily saying we HAVE to use that here, but I don't think media type is accurate.
- [ ] WIP - I think you have the video title and duration in the wrong place.
- [ ] Why does the type size change here? CTA block should be using expressive-heading-05 for the heading, and heading-03 for that copy.

@kirbyevans 

- [ ] Content section heading in the left column area should only span 3 columns

- [ ] Main content area needs shifted left to align with the 16-column grid.

- [ ] Appears to be some alignment issues with the Content item component. Additional space getting added to title on shorter Content items. @asudoh Any guidance on this?

- [ ] CTA text link with icon should consistently align based on longest content item character count. See design spec for more detail. @asudoh Any guidance on this?

- [ ] WIP - I want to also confirm that an additional CTA text link with icon can be added to a Content item for a max of two links if needed.

- [ ] WIP - Bottom spacing for items in the last row should be 160px.

**Comment by asudoh** [comment]
> Appears to be some alignment issues with the Content item component. Additional space getting added to title on shorter Content items. @asudoh Any guidance on this?

Some quick findings are:

* We went away from using `bx--grid` for container `<div>s` in content pattern stories, including [one for CTA block](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3cfee32aefc3015eeeacf74523c435b015b12433/packages/web-components/src/components/cta-block/__stories__/cta-block.stories.ts#L167). That should be replaced with `dds-ce-demo-devenv--simple-grid`.
* Along with `dds-ce-demo-devenv--simple-grid`, we define how much grid columns the story content takes. You can see it at [here](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.16.1/packages/web-components/.storybook/container.scss#L77-L141). Seems that [one for CTA block uses `dds-ce-demo-devenv--simple-grid--content-layout`](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3cfee32aefc3015eeeacf74523c435b015b12433/packages/web-components/src/components/cta-block/__stories__/cta-block.stories.ts#L167), which means [taking 5th to 12th grid columns in `lg` or larger breakpoints](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.16.1/packages/web-components/.storybook/container.scss#L114-L118). Seems that we want the whole 1st to 16th columns for CTA block, and thus `dds-ce-demo-devenv--simple-grid--content-section` is more applicable.

One you apply above two changes (replacing `bx--grid` with `dds-ce-demo-devenv--simple-grid` as well as replacing `dds-ce-demo-devenv--simple-grid--content-layout` with `dds-ce-demo-devenv--simple-grid--content-section`), you'll see that `<dds-cta-block>` now aligns to the Carbon grid well. Now let's take a look at what's inside `<dds-cta-block>`:

* The grid columns inside `<dds-cta-block>` should be 25%/75% split. However, I see [a different split](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3cfee32aefc3015eeeacf74523c435b015b12433/packages/styles/scss/components/cta-block/index.scss#L30) - I guess it's for getting some padding? If so, the padding should be applied by something else.
* The [`column-gap` applied in the grid](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3cfee32aefc3015eeeacf74523c435b015b12433/packages/styles/scss/components/cta-block/index.scss#L27) seems to make the subgrid of `<div class="bx--content-section__children">` off from the parent grid. I recommend removing the gutter there and use different mean to add padding to the heading/copy/link above the cards.
* `<div class="bx--cta-block__items">` has [negative margin adjustment](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3cfee32aefc3015eeeacf74523c435b015b12433/packages/styles/scss/components/cta-block/index.scss#L43-L44). It has adverse side effect and we are moving away from it. Removing such negative margin along with other fixes ensures correct grid alignment.

> CTA text link with icon should consistently align based on longest content item character count. See design spec for more detail. @asudoh Any guidance on this?

If it means the vertical position of the CTA text link, the best approach is ensuring that the CTA text link is placed at the bottom of `<dds-cta-block-item>`.

**Comment by rastogir** [comment]
Thanks @asudoh for the detailed guidance! I cleaned up the classes and was able to achieve vertical position of the CTA text link in `<dds-cta-block-item>` via flex properties. Attached a screenshot. CTA block is going behind the feature flag because of some refactor work. Do I need to delete CodeSandBox example? 
Screen Shot 2021-03-23 at 4 04 54 PM

@andysherman2121 @photodow @RaphaelAmadeu 

As discussed CTA block is behind the feature flag and CTA Section > Cta Block variant is removed. Below are CTA Section and CTA Block refactor details we discussed which can be used for the new ticket.

## CTA Section > Default variant will become CTA Block. 
Screen Shot 2021-03-23 at 5 08 39 PM

  - Under the hood CTA Block should utilize Content Block component
     - Use `<dds-content-block-heading>` for CTA Block heading. This is implemented correctly in CTA Section.
     - Use `<dds-content-block-paragraph>` or `<dds-content-block-copy>` for CTA Block copy. Currently `<dds-cta-section-item-copy>` is being used which extend `DDSContentItemCopy`. This will need to be corrected.
     - Currently Content Block component doesn't have a CTA. Functional requirement is to create one and use it for CTA Block CTA
     - This newly created Content Block CTA should also be used in other content block components such Content Block Cards, Content Block Mixed, Content Block Media, Content Block Segmented, Content Block Simple. Technical implementation for this refactor needs to be discussed.
     - CTA Block CTA should have a prop to switch between button group, text link and none.  
   - CTA Block will have below variants
   
image (16)

    - Default
    
Screen Shot 2021-03-23 at 5 08 39 PM

    - With Content Items
    
Screen Shot 2021-03-23 at 5 14 01 PM

    - With Link List
    
Screen Shot 2021-03-23 at 5 14 07 PM

## CTA Block will become CTA Section.
  - CTA Section variants are
image (16)

<pre>- Default. This will utilize CTA Block > default under the hood. Users should be able to switch between button group, text links and none via a prop.</pre>
  
Screen Shot 2021-03-23 at 5 21 46 PM

    - With Content Items. Users should be able to switch image to video, pictogram, statistics and none via a prop. This is already at CTA Block > Default variant which is behind the feature flag. Design spec for this variant is at https://ibm.ent.box.com/s/shr0pnhof0yzk2b5lh0nokwopgxxb9rr
   
Screen Shot 2021-03-23 at 5 24 37 PM

    - With Link List. 
Screen Shot 2021-03-23 at 5 30 47 PM

Refactor work will also include below two feedbacks from this PR:

- Bottom spacing for items in the last row should be 160px from Kirby.
- I think you have the video title and duration in the wrong place from James.

**Comment by asudoh** [comment]
>  CTA block is going behind the feature flag because of some refactor work. Do I need to delete CodeSandBox example?

@jeffchew What is our policy for CodeSandbox examples of roped-off components? Thanks!

**Comment by jeffchew** [comment]
We can exclude if behind a feature flag, but my understanding is that this feature is going straight to production ready.

**Comment by shixiedesign** [comment]
Hi Ravenna, thank you and this is a lot of work! My understanding is this is not a feature-flagged component, and should go into prod. But I noticed your comment:

> CTA block is going behind the feature flag because of some refactor work. Do I need to delete CodeSandBox example?

I suggest if this needs more work, we don't rush it by putting it behind feature flag. If the dev team has agreed to put it behind feature-flag, it would also require a different tag to ensure `experimental` environment gets build. So I cannot review it in the PR preview builds right now. Could you double check that feature flag decision with the dev team? Thank you!

**Comment by rastogir** [comment]
@asudoh @jeffchew @shixiedesign Component is almost production ready. There is a requirement to rename existing CTA Section to CTA Block and address some technical debt which will require significant refactor work. Refactor work is detailed in my last comment. A seasoned DDS developer would be appropriate for the refactor job hence the reason why this component is put behind the feature flag. If needed there is an option to decrease the scope of refactor work. Happy to schedule a meeting to go over the technical details.

**Comment by rastogir** [comment]
@shixiedesign Experimental preview is built and it's ready for your review https://ibmdotcom-web-components-experimental.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5530/index.html?path=/story/components-cta-block--default

**Comment by jeffchew** [comment]
Closing this PR, @RaphaelAmadeu will be opening a new one.

**Pull request closed without merging**